File: app/services/combos.py
"""
Combo registry — named provider+model chains.

A combo aliases a friendly name ("smart", "fast", "coder") to an ordered
list of targets. Requests can use `model="combo:smart"` to route through it.
Strategies:
- strict:      try targets in order, fail over on error
- round_robin: rotate the starting target
- least_used:  prefer the target with the lowest recent request count
- race:        fire several targets in parallel, first valid response wins
"""

import json
import logging
import time
from dataclasses import dataclass, field

from app.database.redis import get_redis

logger = logging.getLogger(__name__)

COMBO_PREFIX = "gw:combo:"
COMBO_INDEX = "gw:combos:index"
COMBO_RR_PREFIX = "gw:combo:rr:"


@dataclass
class ComboTarget:
    """One target in a combo's chain."""
    provider: str
    model: str | None = None  # None → provider default model
    weight: int = 1

    def to_dict(self) -> dict:
        return {"provider": self.provider, "model": self.model, "weight": self.weight}


@dataclass
class Combo:
    """A named routing chain with an execution strategy."""
    name: str
    targets: list[ComboTarget]
    strategy: str = "strict"  # strict | round_robin | least_used | race
    race_size: int = 2        # only used by strategy="race"
    created_at: float = field(default_factory=time.time)

    def to_dict(self) -> dict:
        return {
            "name": self.name,
            "targets": [t.to_dict() for t in self.targets],
            "strategy": self.strategy,
            "race_size": self.race_size,
            "created_at": self.created_at,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Combo":
        return cls(
            name=data["name"],
            targets=[ComboTarget(**t) for t in data.get("targets", [])],
            strategy=data.get("strategy", "strict"),
            race_size=data.get("race_size", 2),
            created_at=data.get("created_at", time.time()),
        )


class ComboStore:
    """Redis-backed combo registry with in-memory fallback."""

    def __init__(self):
        self._local: dict[str, Combo] = {}

    async def list(self) -> list[Combo]:
        try:
            redis = await get_redis()
            keys = await redis.smembers(COMBO_INDEX)
            combos = []
            for raw in keys:
                name = raw.decode() if isinstance(raw, bytes) else raw
                raw_data = await redis.get(f"{COMBO_PREFIX}{name}")
                if raw_data:
                    combos.append(Combo.from_dict(json.loads(raw_data)))
            return combos or list(self._local.values())
        except Exception as e:
            logger.debug(f"Combo list via Redis failed ({e}); using local cache")
            return list(self._local.values())

    async def get(self, name: str) -> Combo | None:
        """Look up a combo by name, with or without the 'combo:' prefix."""
        clean = name.removeprefix("combo:")
        try:
            redis = await get_redis()
            raw = await redis.get(f"{COMBO_PREFIX}{clean}")
            if raw:
                return Combo.from_dict(json.loads(raw))
        except Exception as e:
            logger.debug(f"Combo get via Redis failed ({e})")
        return self._local.get(clean)

    async def save(self, combo: Combo):
        self._local[combo.name] = combo
        try:
            redis = await get_redis()
            await redis.set(f"{COMBO_PREFIX}{combo.name}", json.dumps(combo.to_dict()))
            await redis.sadd(COMBO_INDEX, combo.name)
        except Exception as e:
            logger.warning(f"Failed to persist combo '{combo.name}' to Redis: {e}")

    async def delete(self, name: str) -> bool:
        clean = name.removeprefix("combo:")
        existed_local = self._local.pop(clean, None) is not None
        try:
            redis = await get_redis()
            removed = await redis.delete(f"{COMBO_PREFIX}{clean}")
            await redis.srem(COMBO_INDEX, clean)
            return bool(removed) or existed_local
        except Exception as e:
            logger.warning(f"Failed to delete combo '{clean}' from Redis: {e}")
            return existed_local

    async def next_rotation_index(self, combo_name: str, modulo: int) -> int:
        """Atomic round-robin counter shared across workers."""
        if modulo <= 0:
            return 0
        try:
            redis = await get_redis()
            return int(await redis.incr(f"{COMBO_RR_PREFIX}{combo_name}")) % modulo
        except Exception:
            combo = self._local.get(combo_name)
            if combo is None:
                return 0
            idx = getattr(combo, "_local_rr", 0)
            combo._local_rr = idx + 1
            return idx % modulo


combo_store = ComboStore()

File: gateway/internal/combos/combos.go
// Package combos is the model-alias registry.
//
// A combo is a named list of targets (provider + optional model) plus a
// routing strategy. Requests use model="combo:<name>" to route through it.
// Fresh Persp: stored in Redis with an in-memory mirror.
package combos

import (
	"context"
	"encoding/json"
	"log/slog"
	"strings"
	"sync"
	"time"

	"github.com/redis/go-redis/v9"
)

// Strategy names.
const (
	StrategyStrict     = "strict"
	StrategyRoundRobin = "round_robin"
	StrategyLeastUsed  = "least_used"
	StrategyRace       = "race"
)

// Target names one provider (+optional model override) in a combo.
type Target struct {
	Provider string `json:"provider"`
	Model    string `json:"model,omitempty"`
	Weight   int    `json:"weight,omitempty"`
}

// Combo is the user-facing named chain.
type Combo struct {
	Name      string    `json:"name"`
	Targets   []Target  `json:"targets"`
	Strategy  string    `json:"strategy"`
	RaceSize  int       `json:"race_size"`
	CreatedAt time.Time `json:"created_at"`
}

// Store keeps combos in Redis ("gw:combo:*") with a local read-through mirror.
type Store struct {
	rdb   redis.UniversalClient
	local map[string]*Combo
	mu    sync.RWMutex
}

// New returns an empty Store bound to Redis.
func New(rdb redis.UniversalClient) *Store {
	return &Store{rdb: rdb, local: map[string]*Combo{}}
}

const (
	comboPrefix = "gw:combo:"
	comboIndex  = "gw:combos:index"
)

// Save upserts a combo.
func (s *Store) Save(ctx context.Context, c *Combo) error {
	if c.Name == "" {
		return errInvalid("name is required")
	}
	if len(c.Targets) == 0 {
		return errInvalid("at least one target is required")
	}
	if c.Strategy == "" {
		c.Strategy = StrategyStrict
	}
	if c.RaceSize < 2 {
		c.RaceSize = 2
	}
	if c.CreatedAt.IsZero() {
		c.CreatedAt = time.Now()
	}
	s.mu.Lock()
	s.local[c.Name] = c
	s.mu.Unlock()
	if s.rdb == nil {
		return nil
	}
	buf, err := json.Marshal(c)
	if err != nil {
		return err
	}
	pipe := s.rdb.TxPipeline()
	pipe.Set(ctx, comboPrefix+c.Name, buf, 0)
	pipe.SAdd(ctx, comboIndex, c.Name)
	if _, err := pipe.Exec(ctx); err != nil {
		slog.Warn("combo persist to Redis failed", "name", c.Name, "err", err)
	}
	return nil
}

// Get returns a combo by name (with or without the "combo:" prefix).
func (s *Store) Get(ctx context.Context, name string) (*Combo, bool) {
	name = strings.TrimPrefix(name, "combo:")

	s.mu.RLock()
	if c, ok := s.local[name]; ok {
		s.mu.RUnlock()
		return c, true
	}
	s.mu.RUnlock()
	if s.rdb == nil {
		return nil, false
	}

	raw, err := s.rdb.Get(ctx, comboPrefix+name).Result()
	if err != nil {
		return nil, false
	}
	var c Combo
	if err := json.Unmarshal([]byte(raw), &c); err != nil {
		return nil, false
	}
	s.mu.Lock()
	s.local[name] = &c
	s.mu.Unlock()
	return &c, true
}

// List returns every registered combo.
func (s *Store) List(ctx context.Context) []*Combo {
	if s.rdb == nil {
		s.mu.RLock()
		defer s.mu.RUnlock()
		out := make([]*Combo, 0, len(s.local))
		for _, c := range s.local {
			out = append(out, c)
		}
		return out
	}
	names, err := s.rdb.SMembers(ctx, comboIndex).Result()
	out := []*Combo{}
	if err == nil {
		for _, n := range names {
			raw, err := s.rdb.Get(ctx, comboPrefix+n).Result()
			if err != nil {
				continue
			}
			var c Combo
			if json.Unmarshal([]byte(raw), &c) == nil {
				out = append(out, &c)
			}
		}
		return out
	}
	s.mu.RLock()
	defer s.mu.RUnlock()
	for _, c := range s.local {
		out = append(out, c)
	}
	return out
}

// Delete removes a combo.
func (s *Store) Delete(ctx context.Context, name string) bool {
	name = strings.TrimPrefix(name, "combo:")
	if s.rdb != nil {
		pipe := s.rdb.TxPipeline()
		pipe.Del(ctx, comboPrefix+name)
		pipe.SRem(ctx, comboIndex, name)
		_, _ = pipe.Exec(ctx)
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	_, existed := s.local[name]
	delete(s.local, name)
	return existed
}

func errInvalid(msg string) error { return &invalidComboError{msg: msg} }

type invalidComboError struct{ msg string }

func (e *invalidComboError) Error() string { return e.msg }
